lsp/nls/src/usage.rs

```diff
@@ -115,11 +111,20 @@ impl UsageLookup {
 
     /// Return the definition site of `ident`.
     pub fn def(&self, ident: &LocIdent) -> Option<&DefWithPath> {
-        ident
-            .pos
-            .as_opt_ref()
-            .and_then(|span| self.def_table.get(span))
-            .and_then(|env| env.get(&ident.ident))
+        // First try to look up the definition in our symbols table. If that fails,
+        // find the active environment and look up the ident in it.
+        //
+        // We check the symbols table first so that we can retrieve the definition
+        // in the case that the ident is already pointing straight at it. For example, in
+        // `let x = 3`, the environment containing `x` doesn't define `x` but we still
+        // want `def(x)` to return this definition.
```

Comment:
I think I didn't understand this part. Do you have an example where you do a goto definition or hover, and where NOT looking into the symbol table would be an issue?

Reply:
If the input is `let x = 3 in x` and you hover on either `x`, you want to see that `x` is a `Number`. But if you look up the first `x` in the `def_table` it returns the empty (well, apart from `std`) environment because that's the active environment at that location. (Maybe the `def_table` should really be called the `env_table` because that's actually what it's storing: the environment of each term.)